### PR TITLE
Updates for smoother compilation on ppc64le (SummitDev) architecture.

### PR DESCRIPTION
**Pull request by atillack**
This pull request updates the compiler detection and settings for the newer IBM XL compiler on the ppc64le architecture as used on OLCF SummitDev. Care was taken that legacy settings for AIX are kept around.

Additionally, the build recipe for OLCF SummitDev is updated to reflect the current software stack and this change.

### CODE REVIEW

**Comment by prckent** [comment]
ok to test

File: CMake/IBMCompilers.cmake [hunk 1]
```diff
@@ -54,3 +55,43 @@ IF(QMC_BITS MATCHES 64)
...
+ENDIF()
+
+# Suppress compile warnings
+SET(CMAKE_C_FLAGS   "${CMAKE_C_FLAGS} -Wno-deprecated -Wno-unused-value")
+SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-deprecated -Wno-unused-value")
+
+# Set standard and other optimization settings
+SET(CMAKE_C_FLAGS   "${CMAKE_C_FLAGS} -std=gnu11 -O3")
+SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++11 -qnoxlcompatmacros -O3")
+
```
Comment (ye-luo) [inline CMake/IBMCompilers.cmake]: Do we need to add -std=gnu11 by hand? Does the upper level cmake add -std=c++11 and does it work?
Doesn't the build type release add -O3?

File: CMake/IBMCompilers.cmake [hunk 2]
```diff
@@ -54,3 +55,43 @@ IF(QMC_BITS MATCHES 64)
...
+# Clean up flags
+
+IF(CMAKE_C_FLAGS MATCHES "-qhalt=e")
+  SET(CMAKE_C_FLAGS "")
+ENDIF()
+IF(CMAKE_CXX_FLAGS MATCHES "-qhalt=e")
+  SET(CMAKE_CXX_FLAGS "")
+ENDIF()
+
```
Comment (ye-luo) [inline CMake/IBMCompilers.cmake]: This seems like a hack. Is this needed because cmake doesn't play well with XL?

File: CMakeLists.txt [hunk 3]
```diff
@@ -393,7 +393,12 @@ ELSE(CMAKE_TOOLCHAIN_FILE)
     MESSAGE(WARNING "No compiler identified")
   ENDIF()
   MESSAGE(STATUS "Compiler: ${COMPILER}")
-  MESSAGE(STATUS "Compiler Version: ${CMAKE_CXX_COMPILER_VERSION}")
+  # Workaround for not properly identified CXX compiler in case of IBM XL compiler
+  IF( ${COMPILER} MATCHES "IBM" )
+    MESSAGE(STATUS "Compiler Version: ${CMAKE_C_COMPILER_VERSION}")
+  ELSE()
+    MESSAGE(STATUS "Compiler Version: ${CMAKE_CXX_COMPILER_VERSION}")
+  ENDIF()
```
Comment (ye-luo) [inline CMakeLists.txt]: Is it possible to print inside  CMake/IBMCompilers.cmake

Comment (ye-luo) [inline CMakeLists.txt]: Is it possible to print inside CMake/IBMCompilers.cmake

File: manual/installation.tex [hunk 4]
```diff
@@ -632,10 +610,9 @@ \subsubsection{Building QMCPACK}
 module load cuda
 mkdir build_summitdev
 cd build_summitdev
-cmake -DCMAKE_C_COMPILER="mpicc" \
-      -DCMAKE_C_FLAGS="-std=gnu11 -qsmp -O3" \
-      -DCMAKE_CXX_COMPILER="mpicxx" \
-      -DCMAKE_CXX_FLAGS="-std=gnu++11 -qnoxlcompatmacros -qsmp -O3" \
+cmake -DCMAKE_C_COMPILER="mpixlc" \
+      -DCMAKE_CXX_COMPILER="mpixlC" \
+      -DBUILD_LMYENGINE_INTERFACE=0 \
```
Comment (ye-luo) [inline manual/installation.tex]: BUILD_LMYENGINE_INTERFACE=0 is not necessary. QMC_CUDA=1 will disable it.

File: CMake/IBMCompilers.cmake (hunk 2 above)
Comment (atillack) [inline CMake/IBMCompilers.cmake]: Yes, cmake populates those flags with "-qthreaded -qhalt=e" where -qthreaded is not doing anything and -qhalt=e is not supported anymore on the newer architecture. The other work-around would be to pass -DCMAKE_C_FLAGS="" and -DCMAKE_CXX_FLAGS="" to the cmake command which is not overly user-friendly.

File: CMakeLists.txt (hunk 3 above)
Comment (prckent) [inline CMakeLists.txt]: If CMAKE_C_COMPILER_VERSION not defined with cmake on summitdev and IBM compilers?

Why not print both C and CXX cmake compiler versions all the time? (Note that cmake does not guarantee these will be set, but it is good to see)

File: manual/installation.tex (hunk 4 above)
Comment (atillack) [inline manual/installation.tex]: You are technically correct (the best kind of correct) ;-)

I left it in for those folks who want to compile the CPU version in which case the CXX11 test now passes and formic/utils MPI implementation causes a hard compile error as the C++ bindings are no longer available on SummitDev.

File: CMakeLists.txt (hunk 3 above)
Comment (atillack) [inline CMakeLists.txt]: Sure, but this message is originally output at this location and that's why I left it there. If I leave the original line untouched and add some output to IBMCompilers.cmake we would have two contradicting lines of output...

File: manual/installation.tex (hunk 4 above)
Comment (prckent) [inline manual/installation.tex]: BUILD_LMYENGINE_INTERFACE=0 needed for CPU builds

**Comment by atillack** [comment]
@ye-luo That's a great question (easier CUDA arch passing) and is probably easy to implement - it does seem hard-coded to sm_35 CUDA version >= 6.5 and sm_20 below.

File: manual/installation.tex (hunk 4 above)
Comment (ye-luo) [inline manual/installation.tex]: @prckent Is this because of the MPI in the optimizer library?

File: CMakeLists.txt (hunk 3 above)
Comment (ye-luo) [inline CMakeLists.txt]: I was wrong. DO NOT move it in CMake/IBMCompilers.cmake.
Somehow we already have printout in the very beginning, for example.
 -- The C compiler identification is GNU 6.2.0
-- The CXX compiler identification is GNU 6.2.0
Here we print again.
-- Compiler: GNU
-- Compiler Version: 6.2.0
Unless they can be different, we can just remove the printout here.

File: CMake/IBMCompilers.cmake (hunk 2 above)
Comment (ye-luo) [inline CMake/IBMCompilers.cmake]: Adding CMAKE_C_FLAGS and CMAKE_CXX_FLAGS is definitely not desired.

File: manual/installation.tex (hunk 4 above)
Comment (ye-luo) [inline manual/installation.tex]: @atillack Clear. Thanks.

File: CMake/IBMCompilers.cmake (hunk 1 above)
Comment (atillack) [inline CMake/IBMCompilers.cmake]: Interestingly, gnu11 and gnu++11 used to be needed but this does not seem to be the case anymore - c++11 works fine and gets added automatically when needed, unlike -O3 which is not automatically added for me. I'll make another commit after some more testing.

Comment (atillack) [inline CMake/IBMCompilers.cmake]: Well, never mind, the flags are needed as they are in this commit. While it compiles fine with c++11, I could not successfully run the created binary (tested with the GPU version) ...

Comment (ye-luo) [inline CMake/IBMCompilers.cmake]: Could you paste
-- CXX compiler flags:
-- C compiler flags:
reported by cmake?

**Comment by ye-luo** [comment]
Can we drop sm_20(fermi) completely? We don't even have any machine to test this case.

**Comment by prckent** [comment]
Yes, we can drop Fermi. Large production machines have Kepler or Pascal.

File: CMake/IBMCompilers.cmake (hunk 1 above)
Comment (atillack) [inline CMake/IBMCompilers.cmake]: Both C and CXX compiler flags going into IBMCompilers.cmake are "-qthreaded -qhalt=e".

**Comment by prckent** [comment]
Any objections to merge this version?

**Comment by ye-luo** [comment]
wait

File: CMakeLists.txt (hunk 3 above)
Comment (ye-luo) [inline CMakeLists.txt]: On a Cray, I got
-- The C compiler identification is Intel 17.0.0.20170213
-- The CXX compiler identification is Intel 17.0.0.20170213
and later
-- Compiler: Intel
-- Compiler Version: 17.0.0.20170213
It is basically redundant info and better just remove.

Comment (ye-luo) [inline CMakeLists.txt]: Any objection to remove?

Comment (atillack) [inline CMakeLists.txt]: Not always though - the first lines of output come from CMake's compiler tests which we have no influence on, while the later lines are our output. I agree that the version info may be redundant but I would urge to at least keep the compiler info as this is our only line of output to verify that our own tests chose wisely.

Comment (ye-luo) [inline CMakeLists.txt]: Please remove the version printout.

**Comment by ye-luo** [comment]
I have added the switch allowing changing CUDA architecture. @atillack Please check and update your build recipe for SummitDev. use -D CUDA_ARCH=sm_60 to overwrite the default sm_35.

**Comment by atillack** [comment]
@ye-luo Thanks! The build still works fine and I'm updating the build recipe and the compiler version output with the next (and hopefully final) commit.

**Comment by ye-luo** [comment]
+1 OK to merge

**Pull request merged**